### PR TITLE
Fix computation of caches with a lot of features in the database

### PR DESCRIPTION
**Pull request by jocel1**
This change is [Reviewable](https://reviewable.io/reviews/prestashop/ps_facetedsearch/32)

### CODE REVIEW

File: ps_facetedsearch.php [hunk 1]
```diff
@@ -762,7 +768,9 @@ private static function indexPrices($cursor = null, $full = false, $ajax = false
             do {
                 $cursor = (int) self::indexPricesUnbreakable((int) $cursor, $full, $smart);
                 $time_elapsed = microtime(true) - $start_time;
-            } while ($cursor < $nb_products && Tools::getMemoryLimit() > memory_get_peak_usage() && $time_elapsed < $max_executiontime);
+            } while ($cursor < $nb_products
+            && (Tools::getMemoryLimit() == -1 || Tools::getMemoryLimit() > memory_get_peak_usage())
+            && $time_elapsed < $max_executiontime);
```
Comment (eternoendless) [inline ps_facetedsearch.php]: Small indentation problem here, should be:
```
while($stmt1
   && $stmt2
);

File: ps_facetedsearch.php [hunk 2]
```diff
@@ -2644,44 +2652,50 @@ public function buildLayeredCategories()
...
+                                    $sqlInsert .= '('.(int) $id_category.', '.(int) $id_shop.', '.(int) str_replace('layered_selection_ag_', '', $key).',
 										\'id_attribute_group\','.(int) $n.', '.(int) $limit.', '.(int) $type.'),';
                                 } elseif (substr($key, 0, 23) == 'layered_selection_feat_') {
-                                    $sql_to_insert .= '('.(int) $id_category.', '.(int) $id_shop.', '.(int) str_replace('layered_selection_feat_', '', $key).',
+                                    $sqlInsert .= '('.(int) $id_category.', '.(int) $id_shop.', '.(int) str_replace('layered_selection_feat_', '', $key).',
 										\'id_feature\','.(int) $n.', '.(int) $limit.', '.(int) $type.'),';
                                 }
+
+                                $nbSqlValuesToInsert++;
+                                if ($nbSqlValuesToInsert > 100) {
```
Comment (eternoendless) [inline ps_facetedsearch.php]: Just FYI this will execute 101 inserts instead of 100 :)

**Comment by marionf** [comment]
@jocel1 

Before I had only one model (with 8 categories checked)

capture du 2018-03-28 17-10-13

Now, I have 8 models (with 1 category checked for each):

capture du 2018-03-28 17-10-02

File: ps_facetedsearch.php [hunk 3]
```diff
@@ -104,11 +104,22 @@ public function install()
...
+                    $categoriesId[] = $category['id_category'];
+                    $i++;
+                    if ($i == 100) {
+                        $this->rebuildLayeredCache(array(), $categoriesId, false);
+                        $categoriesId = array();
+                        $i = 0;
+                    }
+                }
+                if (!empty($categoriesId)) {
+                    $this->rebuildLayeredCache(array(), $categoriesId, false);
```
Comment (eternoendless) [inline ps_facetedsearch.php]: How about replacing the foreach with this?
```php
// get all id_category
$categoryIds = array_column($categories, 'id_category');

// group by 100 ids
$chunks = array_chunk($categories, 100);

// rebuild layered cache for each chunk
foreach ($chunks as $chunk) {
    $this->rebuildLayeredCache(array(), $chunk, false);
}
```

**Comment by jocel1** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/ps_facetedsearch/32)*

**Comment by eternoendless** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/prestashop/ps_facetedsearch/32#-:-LArPjoH3ZSAKCkbSp0Z:bnfp4nl)*

**Comment by eternoendless** [comment]
ping @marionf

**Comment by marionf** [comment]
Hello @jocel1 

I have an exception, when I try to configure the module

capture du 2018-04-26 10-31-11

**Comment by jocel1** [comment]
Hi @marionf ,

You need to make sure https://github.com/PrestaShop/PrestaShop/pull/8962 is properly applied on your branch (if you are on the latest develop version it should be fine)

**Comment by marionf** [comment]
All right thank you @jocel1

**Comment by eternoendless** [comment]
Thank you @jocel1

**Pull request merged**